File: ts/src/day12.ts
import { DayBase } from './day-runner.js';

export class Day12 extends DayBase {
    constructor() {
        super(12);
    }

    public async run(): Promise<void> {
        const lines = await this.getInputLines();
        const caves = new Map<string, string[]>();
        lines.forEach(line => {
            const [left, right] = line.trim().split('-');
            caves.set(left, (caves.get(left) || []).concat(right));
            caves.set(right, (caves.get(right) || []).concat(left));
        });

        console.info(`Day ${this.day}`);
        console.info(`  Part 1:`, this.findPaths(caves));
        console.info(`  Part 2:`, this.findPaths(caves, true));
        console.info('');
    }

    private findPaths(caves: Map<string, string[]>, allowSecondVisit: boolean = false): number {
        let paths = new Set<string>();
        let q = [{ 'key': 'start', 'path': 'start', 'dup': false}];
        while (q.length > 0) {
            let current = q.pop() || { 'key': '', 'path': '', 'dup': false}; // just to keep the TS compiler happy

            (caves.get(current.key) || []).forEach(c => {
                if (c === 'end') {
                    paths.add(current.path);
                } else if ((c[0] >= 'A' && c[0] <= 'Z') || (!current.path.includes(c))) {
                    q.push({ 'key': c, 'path': `${current.path}_${c}`, 'dup': current.dup });
                } else if (allowSecondVisit && c !== 'start' && !current.dup) {
                    q.push({ 'key': c, 'path': `${current.path}_${c}`, 'dup': true });
                }
            });
        }
        return paths.size;
    }
}

File: py/day12.py
import os
from collections import defaultdict

# https://adventofcode.com/2021/day/12

def uniquePathCount(caves: dict[str,list[str]], part: int) -> int:
    paths = set()
    tovisit = [('start','start', False)]
    while tovisit:
        k,path, hasdup = tovisit.pop()

        for c in caves[k]:
            if c == 'end':
                paths.add(path+'_'+c)
            elif c.isupper() or c not in path:
                tovisit.append((c, path+c, hasdup))
            elif part == 2 and not hasdup and c != 'start':
                tovisit.append((c, path+'_'+c, True))
    return len(paths)

caves = defaultdict(lambda: [])
with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day12-input", "r") as f:
    for line in [line.strip().split('-') for line in f.readlines()]:
        caves[line[0]].append(line[1])
        caves[line[1]].append(line[0])

print("Day 12")
print("  Part 1:", uniquePathCount(caves, 1))
print("  Part 2:", uniquePathCount(caves, 2))
